### DISCUSSION TITLE
Empty lua table returns EmptyMap since 5.0.1

### DISCUSSION THREAD

**Discussion by ngavrailov** [xetorthio/jedis#3874] [Q&A]
![Screenshot 2024-06-26 at 17 12 28](https://github.com/redis/jedis/assets/11725117/ec629b65-1d88-4480-8df5-4a3c7bd02bd0)
Hello,
What is the point of this change? Currently we are having issues with existing lua scripts, because of difference when the lua table is populated as list and when it is returned empty. Maybe we are not using the output correct expecting lists? Can you suggest how to handle table responses?
Best regards,
Nikolay

**Comment by thachlp** [comment]
Hi @ngavrailov, I checked, and I see that `AGGRESSIVE_ENCODED_OBJECT` was added since jedis5, there are no previous implementing before.
So could you give more information about the issue, such as the input, output ...

**Reply by ngavrailov** [reply]
I am asking what is the point of this part of the implementation. When 'data' is instance of List to return Collection.emptyMap(). Doesn't make sense for me and also caused issues in our code with existing lua scripts when I upgraded to Jedis 5.

Here is one lua script example which starts to break:
`local bingoTickets = {}
...
return bingoTickets`

`List<Object> scriptResult = (List<Object>) runScript(progressOnTicketsScript`

In case there are no 'bingoTickets' and the table doesn't change we hit an issue casting the result to list. The issue is because the empty lua table is returned as Collection.emptyMap() (as I mentioned earlier and show from the jedis code where it comes from). The worst part is that when the table bingoTickets has anything added it is returned as List, so this code works fine. Doesn't make sense for me why jedis has inconsistency with the data types when one table is filled with data as list and when empty. Looks like it is done for other specific case ignoring you can want to return list sometimes.

Can you explain why this change is done in Jedis 5?

**Reply by ngavrailov** [reply]
Just spotted there is a bug opened already:
https://github.com/redis/jedis/issues/3876

**Comment by willianmrs** [comment]
Hi guys I'm having a similar problem so that I will join this thread. 


I am encountering a significant issue after migrating from Jedis version 4.3 to 5.1. My code, which worked fine in 4.3, is now encountering inconsistent and unexpected behavior.

### Code Overview

Here is a simplified version of my Scala code:

```scala
val result = redisCluster.eval(script, keys, arguments)

import java.util
val asList = result.asInstanceOf[util.ArrayList[_]]
val state = asList.get(0).asInstanceOf[String]
val latestEvents = asList.get(1).asInstanceOf[util.ArrayList[_]].asInstanceOf[util.ArrayList[String]]
```

### Lua Script

The Lua script being executed is as follows:

```lua
local nrt_feature = redis.call('HMGET', nrt_key, "value1", "value2")
local nrt_feature_value = nrt_feature[2]

local events_to_aggregate = redis.call("ZRANGEBYSCORE", key, min, max)

return { nrt_feature_value, events_to_aggregate }
```


### Behavior in Jedis 4.3

In Jedis 4.3, the returned data structures were consistent and the Scala code processed them correctly. For example:

- `nrt_feature_value` was a string containing the expected value.
- `events_to_aggregate` was a list of strings.

### Issues in Jedis 5.1

After upgrading to Jedis 5.1, I am encountering the following problems:

1. **Inconsistent `nrt_feature_value`**: 
   - When `events_to_aggregate` contains elements, `nrt_feature_value` is empty.
   - Conversely, when `events_to_aggregate` is empty, `nrt_feature_value` is populated correctly.

2. **Empty `events_to_aggregate`**:
   - When `events_to_aggregate` is supposed to be empty, it is returned as an empty map (`java.util.Collections$EmptyMap`) instead of an empty list. This is causing `ClassCastException` when attempting to cast to `ArrayList`.

### Example Scenarios

- **Scenario 1**:
  ```lua
  nrt_feature_value = "f_value1"
  events_to_aggregate = ["event1", "event2"]
  ```
  - **Observed Behavior**: In Scala, `events_to_aggregate` can be parsed as a list, but `nrt_feature_value` is empty.

- **Scenario 2**:
  ```lua
  nrt_feature_value = "f_value1"
  events_to_aggregate = []
  ```
  - **Observed Behavior**: `nrt_feature_value` is correctly populated, but `events_to_aggregate` is an empty map, causing type-casting issues.

### Additional Information

It appears that the return type handling of the `eval` command has changed between versions, causing these inconsistencies. This behavior is making it challenging to migrate to the newer version of Jedis.

Could you please advise on whether this is an intended change, or if there might be a bug or configuration issue that I need to address?

Thank you for your assistance!

---

### Details
- **Jedis Version**: 5.1.3
- **Redis Version**: 6.2 (But we have the same behavior on 7.2) 
- **Scala Version**: 2.12
- **Java Version**: 8
- **Sark Version**: 3.4.1 (that happens outside spark, but since it adds a lot of dependency with it, I'm adding the version we are using here)
